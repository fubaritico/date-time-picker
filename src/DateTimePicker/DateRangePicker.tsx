import { CommonPickerProps } from '@types'

import Picker from './components/Picker'

import type { FC } from 'react'

export type DateRangePickerProps = CommonPickerProps<'DATERANGE'>

const DateRangePicker: FC<DateRangePickerProps> = ({
  dateRange,
  onDateRangeChange,
  locale,
  timezone,
  ...rest
}) => {
  return (
    <Picker<'DATERANGE'>
      {...rest}
      dateRange={dateRange}
      onDateRangeChange={onDateRangeChange}
      locale={locale}
      timezone={timezone}
    />
  )
}

export default DateRangePicker
