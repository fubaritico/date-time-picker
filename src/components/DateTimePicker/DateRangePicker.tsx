import Picker from './components/Picker'

import type { CommonPickerProps } from './types'
import type { FC } from 'react'

export type DateRangePickerProps = CommonPickerProps<'DATERANGE'>

const DateRangePicker: FC<DateRangePickerProps> = ({
  dateRange,
  onDateRangeChange,
  locale,
  timezone,
  ...rest
}) => {
  return (
    <Picker<'DATERANGE'>
      {...rest}
      dateRange={dateRange}
      onDateRangeChange={onDateRangeChange}
      locale={locale}
      timezone={timezone}
    />
  )
}

export default DateRangePicker
