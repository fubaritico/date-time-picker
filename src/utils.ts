import { LocaleAwareFormat } from './I18nDate'

import type { ChangeEvent, KeyboardEvent } from 'react'

/**
 * Debounce function that will delay the execution of the passed function
 *
 * @param func - any function
 * @param {number} wait - delay applied to the function
 */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export const debounce = <Fn extends (...args: never[]) => void>(
  func: Fn,
  wait: number
): ((...args: Parameters<Fn>) => void) => {
  let timeoutId: ReturnType<typeof setTimeout>

  return function (...args: Parameters<Fn>) {
    clearTimeout(timeoutId)
    timeoutId = setTimeout(() => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      func.apply(this, args)
    }, wait)
  }
}

/**
 * Handle key down event of 'enter' & 'space' keys for tags that need
 * to be clickable with keyboard due to accessibility reasons
 * Works when an element is focused.
 *
 * @param callback
 */
export const handleKeyDown =
  <Element>(callback: (e?: Event) => void) =>
  (event: KeyboardEvent<Element>): void => {
    if (event.key === 'Enter' || event.key === ' ') {
      // The passed callback has, most of the time, no event argument
      callback(event as never)
    }
  }

/**
 * Function to create a custom change event.
 * @param eventValue
 */
export const createCustomChangeEvent = (
  eventValue: string
): ChangeEvent<HTMLInputElement> => {
  const event = new Event('change', {
    bubbles: true,
    cancelable: true,
  }) as unknown as ChangeEvent<HTMLInputElement>
  Object.defineProperty(event, 'target', {
    writable: true,
    value: { value: eventValue } as HTMLInputElement,
  })
  return event
}

/**
 * Will return options for the Intl.DateTimeFormat constructor based on a format and timezone
 *
 * Important note: the returned option object used for formatting is supposed to mimic LocaleAwareFormat
 * used in momentJS, the goal is to avoid changing the signature (for now) of the method used in I18nDate if possible.
 *
 * @see: https://tc39.es/proposal-intl-datetime-style/
 *
 *    interface DateTimeFormatOptions {
 *         calendar?: string | undefined;
 *         dayPeriod?: "narrow" | "short" | "long" | undefined;
 *         numberingSystem?: string | undefined;
 *
 *         dateStyle?: "full" | "long" | "medium" | "short" | undefined;
 *         timeStyle?: "full" | "long" | "medium" | "short" | undefined;
 *         hourCycle?: "h11" | "h12" | "h23" | "h24" | undefined;
 *     }
 *
 *     interface DateTimeFormatOptions {
 *         localeMatcher?: "best fit" | "lookup" | undefined;
 *         weekday?: "long" | "short" | "narrow" | undefined;
 *         era?: "long" | "short" | "narrow" | undefined;
 *         year?: "numeric" | "2-digit" | undefined;
 *         month?: "numeric" | "2-digit" | "long" | "short" | "narrow" | undefined;
 *         day?: "numeric" | "2-digit" | undefined;
 *         hour?: "numeric" | "2-digit" | undefined;
 *         minute?: "numeric" | "2-digit" | undefined;
 *         second?: "numeric" | "2-digit" | undefined;
 *         timeZoneName?: "short" | "long" | "shortOffset" | "longOffset" | "shortGeneric" | "longGeneric" | undefined;
 *         formatMatcher?: "best fit" | "basic" | undefined;
 *         hour12?: boolean | undefined;
 *         timeZone?: string | undefined;
 *     }
 *
 * @param pLocaleAwareFormat
 * @param timeZone
 */
const getLocaleAwareFormat = (
  pLocaleAwareFormat: LocaleAwareFormat,
  timeZone?: Timezone
): Intl.DateTimeFormatOptions | undefined => {
  switch (pLocaleAwareFormat) {
    case 'l':
      return {
        month: '2-digit',
        day: '2-digit',
        year: 'numeric',
      }
    case 'll':
      return {
        dateStyle: 'medium',
      }
    case 'lll • t':
    case 'lll':
      return {
        dateStyle: 'medium',
        timeStyle: 'short',
      }
    case 'llll':
      return {
        weekday: 'short',
        month: 'short',
        day: 'numeric',
        year: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
      }
    case 'L':
      return {
        month: '2-digit',
        day: '2-digit',
        year: 'numeric',
      }
    case 'LL':
      return {
        month: 'long',
        day: 'numeric',
        year: 'numeric',
      }
    case 'LLL':
      return {
        month: 'long',
        day: 'numeric',
        year: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
      }
    case 'LLLL':
      return {
        weekday: 'long',
        month: 'long',
        day: 'numeric',
        year: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
      }
    case 'L LT':
      return {
        year: 'numeric',
        month: 'numeric',
        day: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
      }
    case 'L LT Z':
      return {
        dateStyle: 'short',
        timeStyle: 'long',
        timeZone,
      }
    case 'LT':
      return {
        hour: 'numeric',
        minute: 'numeric',
      }
    case 'LTS':
      return {
        hour: 'numeric',
        minute: 'numeric',
        second: 'numeric',
      }
    default:
      return undefined
  }
}

/**
 * Locale aware formats
 * --------------------
 * Locale aware date and time formats are also available using LT LTS L LL LLL LLLL.
 * They were added in version 2.2.1, except LTS which was added 2.8.4.
 *
 * Tokens are case-sensitive.
 *
 * Input	Example	Description
 * L  	09/04/1986	Date (in local format)
 * LL 	September 4 1986	Month name, day of month, year
 * LLL	September 4 1986 8:30 PM	Month name, day of month, year, time
 * LLLL	Thursday, September 4 1986 8:30 PM	Day of week, month name, day of month, year, time
 * LT 	8:30 PM	Time (without seconds)
 * LTS	8:30:00 PM	Time (with seconds)
 *
 * See also: https://momentjs.com/docs/#/parsing/string/ (Supported ISO 8601 strings)
 * and the RFC 2822 date time format
 *
 * All country codes can be retrieved using method moment.tz.countries()
 *
 * @param {number} pValue - date under the shape of a unix timestamp UTC
 * @param {string} pLocale - locale as a string under the shape of 'fr_FR'
 * @param {string} pLocaleAwareFormat - format to apply to the date in a locale way
 * @param {string} pTimezone - time to add/substract the hour offset, default
 */
export const formatToLocaleAwareFormat = (
  pValue: number,
  pLocale: string,
  pLocaleAwareFormat: LocaleAwareFormat,
  pTimezone?: Timezone
): string => {
  return (
    new Intl.DateTimeFormat(pLocale.replace('_', '-'), {
      ...getLocaleAwareFormat(pLocaleAwareFormat, pTimezone),
    })
      .format(new Date(pValue))
      // Removing à/at from the formatted date
      .replace(/\s(at|à)/, () => {
        return pLocaleAwareFormat === 'lll • t' ? ' •' : ''
      })
      // Adapting timezone to the format +1:00
      .replace(/\sGMT(-|\+)(\d+)/, (_match: string, p1: string, p2: string) => {
        return ` ${p1}${p2}:00`
      })
  )
}
