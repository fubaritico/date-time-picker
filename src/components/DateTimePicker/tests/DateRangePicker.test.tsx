import { screen, waitFor, within } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import MockDate from 'mockdate'

import DateRangePicker from '../DateRangePicker'

import { setupUncontrolledPicker } from './utils'

import type { AnyPickerComponent } from '../types'

const spyOnDateChangeFn = jest.fn()

const runTests = () => {
  const fixedDate = 1742052493000

  beforeEach(() => {
    MockDate.set(fixedDate) // some arbitrary date
  })

  afterEach(() => {
    MockDate.reset()
  })

  it('should render', async () => {
    const {
      render: { container },
    } = setupUncontrolledPicker(
      fixedDate,
      DateRangePicker as AnyPickerComponent
    )

    await waitFor(() => {
      expect(container).toBeInTheDocument()
    })

    expect(container).toMatchSnapshot()
  })

  /**
   * When uncontrolled, the default value should be that of today.
   */
  describe('Uncontrolled - init', () => {
    // There's a behavior issue with @mona-health/react-input-mask
    // the default value should be the mask (____/__/__), not an empty value
    it('should not pass any value to the inputs on init', () => {
      setupUncontrolledPicker(fixedDate, DateRangePicker as AnyPickerComponent)

      expect(screen.getByTestId('start-input')).toHaveValue('')
      expect(screen.getByTestId('end-input')).toHaveValue('')
    })

    it('should open the date range panel with the start date input only', async () => {
      setupUncontrolledPicker(fixedDate, DateRangePicker as AnyPickerComponent)

      await userEvent.click(
        within(screen.getByTestId('start-input-control')).getByRole('button')
      )

      expect(screen.getByTestId('date-range-panel')).toBeInTheDocument()

      await userEvent.click(document.body)

      await waitFor(() => {
        expect(screen.queryByTestId('date-range-panel')).not.toBeInTheDocument()
      })

      await userEvent.click(
        within(screen.getByTestId('end-input-control')).getByRole('button')
      )

      expect(screen.queryByTestId('date-range-panel')).not.toBeInTheDocument()
    })

    it('should give the opener button the "Choose Date Range" label', () => {
      setupUncontrolledPicker(fixedDate, DateRangePicker as AnyPickerComponent)

      expect(
        within(screen.getByTestId('start-input-control')).getByRole('button')
      ).toHaveAccessibleName('Choose Date Range')
    })

    it('should show the current date in the left panel by default', async () => {
      const { todayTimestamp } = setupUncontrolledPicker(
        fixedDate,
        DateRangePicker as AnyPickerComponent
      )

      await userEvent.click(screen.getByLabelText('Choose Date Range'))

      expect(screen.getByTestId('date-range-panel')).toBeInTheDocument()

      expect(
        within(screen.getByTestId('start-date-panel')).getByTestId(
          todayTimestamp.toString()
        )
      ).toHaveClass('DaysGridCell today blue md')
    })

    it('should open the panel with proper state', async () => {
      setupUncontrolledPicker(fixedDate, DateRangePicker as AnyPickerComponent)

      await userEvent.click(screen.getByLabelText('Choose Date Range'))

      expect(screen.getByTestId('date-range-panel')).toBeInTheDocument()

      expect(
        within(screen.getByTestId('start-date-panel')).getByLabelText(
          'Previous Month'
        )
      ).toBeEnabled()

      expect(
        within(screen.getByTestId('start-date-panel')).getByText('March')
      ).toBeInTheDocument()

      expect(
        within(screen.getByTestId('start-date-panel')).getByText('2025')
      ).toBeInTheDocument()

      expect(
        within(screen.getByTestId('start-date-panel')).getByLabelText(
          'Next Month'
        )
      ).toBeDisabled()

      expect(
        within(screen.getByTestId('end-date-panel')).getByLabelText(
          'Previous Month'
        )
      ).toBeDisabled()

      expect(
        within(screen.getByTestId('end-date-panel')).getByText('April')
      ).toBeInTheDocument()

      expect(
        within(screen.getByTestId('end-date-panel')).getByText('2025')
      ).toBeInTheDocument()

      expect(
        within(screen.getByTestId('end-date-panel')).getByLabelText(
          'Next Month'
        )
      ).toBeEnabled()
    })
  })

  describe('Uncontrolled - interaction', () => {
    it(
      'should enable next button in start date panel and previous button in end date panel ' +
        'if months in panels are not successive',
      async () => {
        setupUncontrolledPicker(
          fixedDate,
          DateRangePicker as AnyPickerComponent
        )

        await userEvent.click(screen.getByLabelText('Choose Date Range'))

        await userEvent.click(
          within(screen.getByTestId('start-date-panel')).getByLabelText(
            'Previous Month'
          )
        )

        expect(
          within(screen.getByTestId('start-date-panel')).getByLabelText(
            'Next Month'
          )
        ).toBeEnabled()

        expect(
          within(screen.getByTestId('end-date-panel')).getByLabelText(
            'Previous Month'
          )
        ).toBeEnabled()
      }
    )

    it('should update month in the panel header when click on previous or next button', async () => {
      setupUncontrolledPicker(fixedDate, DateRangePicker as AnyPickerComponent)

      await userEvent.click(screen.getByLabelText('Choose Date Range'))

      await userEvent.click(
        within(screen.getByTestId('start-date-panel')).getByLabelText(
          'Previous Month'
        )
      )

      expect(
        within(screen.getByTestId('start-date-panel')).getByText('February')
      ).toBeInTheDocument()

      await userEvent.click(
        within(screen.getByTestId('start-date-panel')).getByLabelText(
          'Next Month'
        )
      )

      expect(
        within(screen.getByTestId('start-date-panel')).getByText('March')
      ).toBeInTheDocument()
    })
  })

  describe('Controlled', () => {
    beforeEach(() => {
      spyOnDateChangeFn.mockReset()
    })

    // date used for this set of test: 2024-08-09T16:28:13.000Z
    // const defaultProperties: AnyPickerProps = {
    //   date: 1723220893000,
    //   locale: 'en_US',
    //   pickerMode: 'DATETIME',
    //   timezone,
    // }

    it('should test', async () => {})
  })
}

/**
 * The Date management of the component is not based on moment.js anymore.
 * JS Date object and Intl.DateTimeFormat are used instead.
 * The first set of tests will focus on the basic behavior of the component.
 * And the machine timezone is used (fr, GMT + 1:00)
 *
 * For the sake of reliability, MockDate will be used in those tests to keep them deterministic.
 *
 * A few tests are added to ensure that a provided timezone is taken into account.
 */
describe('DateRangePicker', () => {
  runTests()
})
