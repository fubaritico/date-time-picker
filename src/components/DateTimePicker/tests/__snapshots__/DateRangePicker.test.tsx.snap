// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangePicker should render 1`] = `
<div>
  <div
    style="position: relative;"
  >
    <div
      style="position: relative;"
    >
      <div>
        <div
          class="DateRangeInput blue"
        >
          <div
            data-focus="false"
          >
            <div
              class="HelperText"
            >
              <div
                data-test="start-input-control"
                style="position: relative; width: 100%;"
              >
                <input
                  class="TextField blue md icon-left show-icon hide-focus"
                  data-test="start-input"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <button
                  aria-label="Choose Date Range"
                  class="icon-button"
                  tabindex="0"
                >
                  <svg
                    aria-hidden="true"
                    class="icon"
                    data-jest-file-name="HiMiniCalendarDays.svg"
                    data-jest-svg-name="HiMiniCalendarDays"
                    data-testid="HiMiniCalendarDays"
                  />
                </button>
              </div>
            </div>
            <div
              class="HelperText"
            >
              <div
                data-test="end-input-control"
                style="position: relative; width: 100%;"
              >
                <input
                  class="TextField blue md icon-left show-icon hide-focus"
                  data-test="end-input"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <button
                  class="icon-button"
                  disabled=""
                  tabindex="-1"
                >
                  <svg
                    aria-hidden="true"
                    class="icon"
                    data-jest-file-name="HiMiniCalendarDays.svg"
                    data-jest-svg-name="HiMiniCalendarDays"
                    data-testid="HiMiniCalendarDays"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
